liquidhaskell-boot/src-ghc/Liquid/GHC/API/Extra.hs

```diff
@@ -170,46 +170,43 @@ apiCommentsParsedSource ps =
     spanToLineColumn =
       fmap (\s -> (srcSpanStartLine s, srcSpanStartCol s)) . srcSpanToRealSrcSpan
 
--- | Adds NOINLINE pragmas to all local bindings in the module.
+-- | Adds NOINLINE pragmas to all bindings in the module.
 --
 -- This prevents the simple optimizer from inlining such bindings which might
 -- have specs that would otherwise be left dangling.
 --
 -- https://gitlab.haskell.org/ghc/ghc/-/issues/24386
 --
-addNoInlinePragmasToLocalBinds :: ParsedModule -> ParsedModule
-addNoInlinePragmasToLocalBinds ps =
-    ps { pm_parsed_source = go (pm_parsed_source ps) }
+addNoInlinePragmasToBinds :: TcGblEnv -> TcGblEnv
+addNoInlinePragmasToBinds tcg = tcg{ tcg_binds = go False (tcg_binds tcg) }
   where
-    go :: forall a. Data a => a -> a
-    go = gmapT ((id `extT` addNoInlinePragmas) . go)
-
-    addNoInlinePragmas :: HsValBinds GhcPs -> HsValBinds GhcPs
-    addNoInlinePragmas = \case
-      ValBinds x binds sigs ->
-          ValBinds x binds (newSigs ++ dropInlinePragmas sigs)
-        where
-          dropInlinePragmas = filter (not . isInlineSig . unLoc)
-
-          isInlineSig InlineSig{} = True
-          isInlineSig _ = False
-
-          newSigs = concatMap (traverse noInlinePragmaForBind) $ bagToList binds
-
-          noInlinePragmaForBind :: HsBindLR GhcPs GhcPs -> [Sig GhcPs]
-          noInlinePragmaForBind bndr =
-              [ InlineSig
-                  []
-                  (noLocA b)
-                  defaultInlinePragma
-                    { inl_inline = NoInline $ SourceText $ occNameFS $ rdrNameOcc b
-                    , inl_act = NeverActive
-                    }
-              | b <- collectHsBindBinders CollNoDictBinders bndr
-              ]
-
-      bnds -> bnds
-
+    go :: forall a. Data a => Bool -> a -> a
+    go isAbs = gmapT $ go isAbs `extT` markHsBind isAbs
+
+    markHsBind :: Bool -> HsBind GhcTc -> HsBind GhcTc
+    markHsBind isAbs = \case
+        VarBind ext var rhs -> VarBind ext (markId isAbs var) (go False rhs)
+        FunBind ext var matches -> FunBind ext (markId isAbs <$> var) (go False matches)
+        PatBind ext lhs mult rhs -> PatBind ext lhs mult (go False rhs)
+        PatSynBind{} -> error "markNoInline: PatSynBind"
+        XHsBindsLR absBind@AbsBinds{ abs_binds = binds
+                                   , abs_exports = abes } -> XHsBindsLR $ absBind
+          { abs_binds = go True binds
+          , abs_exports = map markABE abes }
+
+    markABE :: ABExport -> ABExport
+    markABE abe@ABE{ abe_poly = poly
+                   , abe_mono = mono } = abe
+      { abe_poly = markId False poly
+      , abe_mono = markId False mono }
+
+    markId :: Bool -> Id -> Id
+    markId isAbs var
+      | isAbs
+      = var
```

Comment:
I think I'd need some comment about why ids in `AbsBinds` need to be left alone.

Reply:
"Because that is what matches the old behaviour" is the only answer I can give you. You will have to answer that yourself -- why did you patch the module before typechecking exactly the way you did it?

Reply:
> why did you patch the module before typechecking?

It looked equally good to do it before or after typechecking, no other reason.

If there is no special motivation to skip ids in AbsBinds, then I'd say they are better marked as the rest.

Reply:
No no, I think you misunderstand.

I've put in a whole lot of work into making the desugarer output match what we had before. Everything in this patch is carefully orchestrated to . You HAVE to leave `AbsBind` `Id`s alone for that to work. If you make them non-inlined, what you get out of the desugarer is a garbled mess that LH can't make heads or tails of.

Reply:
Oh, I pretty much can misunderstand everything in LH and the GHC API. When it comes to motivating this code, I think it is better to put it explicitly in comments, or other contributors and myself won't know why things are done the way they are.